File: Problems & Solutions/1833. Maximum Ice Cream Bars/06 January 2022/Approach 2 - Counting Sort (Greedy)/Maximum Ice Cream Bars.py
class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        n, icecreams = len(costs), 0
        m = max(costs)

        costsFrequency = [0] * (m + 1)
        for cost in costs:
            costsFrequency[cost] += 1

        for cost in range(1, m + 1):
            if not costsFrequency[cost]:
                continue
            if coins < cost:
                break
            
            count = min(costsFrequency[cost], coins // cost)
            coins -= cost * count
            icecreams += count
            
        return icecreams

File: Problems & Solutions/1833. Maximum Ice Cream Bars/06 January 2022/Approach 2 - Counting Sort (Greedy)/Maximum Ice Cream Bars.swift
class Solution {
    func maxIceCream(_ costs: [Int], _ totalCoins: Int) -> Int {
        let n: Int = costs.count
        var icecreams: Int = 0
        var coins: Int = totalCoins
        let m: Int = costs.max()!

        var costsFrequency: [Int] = [Int](repeating: 0, count: (m + 1))
        for cost in costs {
            costsFrequency[cost] += 1
        }

        for cost in 1...m {
            guard costsFrequency[cost] != 0 else {
                continue
            }
            guard coins >= cost else {
                break
            }
            
            let count: Int = min(costsFrequency[cost], coins / cost)
            coins -= cost * count
            icecreams += count
        }

        return icecreams
    }
}